File: mobile/android/app/src/main/java/com/volttracker/obdpoc/DiagnosticsShareIntent.java
package com.volttracker.obdpoc;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.util.Log;
import androidx.core.content.FileProvider;
import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.zip.ZipEntry;
import java.util.zip.ZipOutputStream;

/**
 * Builds a one-zip share intent for the diagnostics share-sheet flow. The zip pulls together:
 *
 * <ul>
 *   <li>the most recent {@value #MAX_SESSION_LOGS} per-session JSONLs from {@code files/obd-logs/}
 *   <li>the rolling app log + its rolled sibling from {@code files/app-log/}
 *   <li>the session summary rollup {@code sessions-summary.jsonl}
 * </ul>
 *
 * <p>The zip is written to {@code cacheDir/diagnostics/} and exposed via the project's {@code
 * FileProvider} — the bridge call wraps the returned intent in {@code Intent.createChooser} and
 * starts it.
 */
final class DiagnosticsShareIntent {

    /** Most recent N per-session JSONLs we include. */
    static final int MAX_SESSION_LOGS = 5;

    private static final String TAG = "DiagnosticsShareIntent";
    private static final String SUBDIR = "diagnostics";
    private static final String SESSION_LOG_DIR = "obd-logs";
    private static final String APP_LOG_DIR = "app-log";
    private static final String SUMMARY_NAME = "sessions-summary.jsonl";

    private DiagnosticsShareIntent() {}

    /**
     * Zips the diagnostics bundle and returns an {@link Intent#ACTION_SEND} intent ready for {@link
     * Intent#createChooser(Intent, CharSequence)}. Returns {@code null} only if the zip couldn't be
     * written (caller logs and surfaces the failure to the UI).
     */
    static Intent buildIntent(Context ctx) {
        if (ctx == null) {
            return null;
        }
        File zip = buildZip(ctx);
        if (zip == null) {
            return null;
        }
        Uri uri;
        try {
            uri = FileProvider.getUriForFile(ctx, ctx.getPackageName() + ".fileprovider", zip);
        } catch (IllegalArgumentException ex) {
            // FileProvider couldn't expose the file — usually a misconfigured file_paths.xml.
            Log.e(TAG, "FileProvider rejected diagnostics zip: " + zip, ex);
            return null;
        }
        Intent intent = new Intent(Intent.ACTION_SEND);
        intent.setType("application/zip");
        intent.putExtra(Intent.EXTRA_STREAM, uri);
        intent.putExtra(Intent.EXTRA_SUBJECT, "VoltTracker diagnostics");
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        return intent;
    }

    /**
     * Builds the zip file (exposed package-private so tests can verify its contents without driving
     * the FileProvider). Returns {@code null} on IO failure.
     */
    static File buildZip(Context ctx) {
        File outDir = new File(ctx.getCacheDir(), SUBDIR);
        if (!outDir.exists() && !outDir.mkdirs()) {
            Log.e(TAG, "could not create diagnostics dir: " + outDir);
            return null;
        }
        // Clear stale zips so the cache doesn't accumulate one per share over the device's
        // lifetime. We keep only the one we're about to build.
        File[] stale = outDir.listFiles();
        if (stale != null) {
            for (File f : stale) {
                if (f.isFile() && !f.delete()) {
                    Log.w(TAG, "could not delete stale diagnostics file: " + f);
                }
            }
        }
        File zipFile = new File(outDir, "diagnostics-" + System.currentTimeMillis() + ".zip");
        File filesDir = ctx.getFilesDir();
        try (FileOutputStream fos = new FileOutputStream(zipFile);
                ZipOutputStream zos = new ZipOutputStream(fos)) {
            addRecentSessionLogs(zos, new File(filesDir, SESSION_LOG_DIR));
            addAppLogs(zos, new File(filesDir, APP_LOG_DIR));
            addSummary(zos, new File(new File(filesDir, SESSION_LOG_DIR), SUMMARY_NAME));
        } catch (IOException ex) {
            Log.e(TAG, "diagnostics zip failed", ex);
            // Best-effort cleanup so callers don't pick up a half-written zip.
            if (zipFile.exists() && !zipFile.delete()) {
                Log.w(TAG, "could not delete partial zip: " + zipFile);
            }
            return null;
        }
        return zipFile;
    }

    private static void addRecentSessionLogs(ZipOutputStream zos, File sessionLogDir)
            throws IOException {
        if (!sessionLogDir.isDirectory()) {
            return;
        }
        // Filter to the per-session JSONLs (session-*.jsonl). The sessions-summary file is also
        // .jsonl but is added separately via addSummary; including it here would either crowd the
        // 5-file budget or — worse — duplicate the entry name in the zip and blow up putNextEntry.
        File[] files =
                sessionLogDir.listFiles(
                        (dir, name) -> name.startsWith("session-") && name.endsWith(".jsonl"));
        if (files == null || files.length == 0) {
            return;
        }
        // Newest first so the slice gives the freshest N files. We use a plain anonymous
        // Comparator with Collections.sort (rather than Comparator.comparingLong / List.sort)
        // because those are API 24+ and the app's minSdk is 23.
        List<File> sorted = new ArrayList<>(Arrays.asList(files));
        Collections.sort(
                sorted,
                new Comparator<File>() {
                    @Override
                    public int compare(File a, File b) {
                        return Long.compare(b.lastModified(), a.lastModified());
                    }
                });
        int limit = Math.min(MAX_SESSION_LOGS, sorted.size());
        for (int i = 0; i < limit; i++) {
            addFile(zos, sorted.get(i), SESSION_LOG_DIR + "/" + sorted.get(i).getName());
        }
    }

    private static void addAppLogs(ZipOutputStream zos, File appLogDir) throws IOException {
        if (!appLogDir.isDirectory()) {
            return;
        }
        File[] files = appLogDir.listFiles();
        if (files == null) {
            return;
        }
        for (File f : files) {
            if (f.isFile()) {
                addFile(zos, f, APP_LOG_DIR + "/" + f.getName());
            }
        }
    }

    private static void addSummary(ZipOutputStream zos, File summary) throws IOException {
        if (summary.isFile()) {
            addFile(zos, summary, SESSION_LOG_DIR + "/" + summary.getName());
        }
    }

    private static void addFile(ZipOutputStream zos, File source, String entryName)
            throws IOException {
        ZipEntry entry = new ZipEntry(entryName);
        entry.setTime(source.lastModified());
        zos.putNextEntry(entry);
        try (BufferedInputStream in = new BufferedInputStream(new FileInputStream(source))) {
            byte[] buf = new byte[8192];
            int n;
            while ((n = in.read(buf)) > 0) {
                zos.write(buf, 0, n);
            }
        }
        zos.closeEntry();
    }
}

File: mobile/android/app/src/main/kotlin/com/volttracker/obdpoc/DiagnosticsShareIntent.kt
package com.volttracker.obdpoc

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import androidx.core.content.FileProvider
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Builds a one-zip share intent for the diagnostics share-sheet flow. The zip pulls together:
 *
 * - the most recent [MAX_SESSION_LOGS] per-session JSONLs from `files/obd-logs/`
 * - the rolling app log + its rolled sibling from `files/app-log/`
 * - the session summary rollup `sessions-summary.jsonl`
 * - `diagnostics-bundle.txt`: a budget-bounded, self-selecting digest (see [DiagnosticsBundle])
 *   sized to hand straight to an external debugging tool without overflowing its context
 *
 * The zip is written to `cacheDir/diagnostics/` and exposed via the project's `FileProvider` -- the
 * bridge call wraps the returned intent in [Intent.createChooser] and starts it.
 */
object DiagnosticsShareIntent {
    /** Most recent N per-session JSONLs we include. */
    const val MAX_SESSION_LOGS = 5

    private const val TAG = "DiagnosticsShareIntent"
    private const val SUBDIR = "diagnostics"
    private const val SESSION_LOG_DIR = "obd-logs"
    private const val APP_LOG_DIR = "app-log"
    private const val SUMMARY_NAME = "sessions-summary.jsonl"

    /**
     * Zips the diagnostics bundle and returns an [Intent.ACTION_SEND] intent ready for
     * [Intent.createChooser]. Returns `null` only if the zip couldn't be written (caller logs and
     * surfaces the failure to the UI).
     */
    @JvmStatic
    fun buildIntent(ctx: Context?): Intent? {
        if (ctx == null) {
            return null
        }
        val zip = buildZip(ctx) ?: return null
        val uri: Uri =
            try {
                FileProvider.getUriForFile(ctx, ctx.packageName + ".fileprovider", zip)
            } catch (ex: IllegalArgumentException) {
                // FileProvider couldn't expose the file -- usually a misconfigured file_paths.xml.
                Log.e(TAG, "FileProvider rejected diagnostics zip: $zip", ex)
                return null
            }
        return Intent(Intent.ACTION_SEND).apply {
            type = "application/zip"
            putExtra(Intent.EXTRA_STREAM, uri)
            putExtra(Intent.EXTRA_SUBJECT, ctx.getString(R.string.share_diagnostics_subject))
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }

    /**
     * Builds an [Intent.ACTION_SEND] intent for the standalone AI digest -- a single redacted,
     * budget-bounded `text/plain` file (see [DiagnosticsBundle]) that can be shared straight into a
     * debugging session without unzipping. Returns `null` if the digest couldn't be written.
     */
    @JvmStatic
    fun buildDigestIntent(ctx: Context?): Intent? {
        if (ctx == null) {
            return null
        }
        val digest = buildDigestFile(ctx) ?: return null
        val uri: Uri =
            try {
                FileProvider.getUriForFile(ctx, ctx.packageName + ".fileprovider", digest)
            } catch (ex: IllegalArgumentException) {
                Log.e(TAG, "FileProvider rejected diagnostics digest: $digest", ex)
                return null
            }
        return Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_STREAM, uri)
            putExtra(Intent.EXTRA_SUBJECT, ctx.getString(R.string.share_diagnostics_subject))
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }

    /**
     * Writes the standalone digest text file (exposed package-private so tests can read it back
     * without driving the FileProvider). Returns `null` on IO failure.
     */
    @JvmStatic
    fun buildDigestFile(ctx: Context): File? {
        val outDir = prepareOutDir(ctx) ?: return null
        val digestFile = File(outDir, "diagnostics-digest-${System.currentTimeMillis()}.txt")
        return try {
            digestFile.writeText(DiagnosticsBundle.build(ctx.filesDir), Charsets.UTF_8)
            digestFile
        } catch (ex: IOException) {
            Log.e(TAG, "diagnostics digest write failed", ex)
            if (digestFile.exists() && !digestFile.delete()) {
                Log.w(TAG, "could not delete partial digest: $digestFile")
            }
            null
        }
    }

    /**
     * Creates the shared `cacheDir/diagnostics/` output dir and clears stale artifacts so the cache
     * keeps only the one we're about to build. Returns `null` if the dir can't be created.
     */
    private fun prepareOutDir(ctx: Context): File? {
        val outDir = File(ctx.cacheDir, SUBDIR)
        if (!outDir.exists() && !outDir.mkdirs()) {
            Log.e(TAG, "could not create diagnostics dir: $outDir")
            return null
        }
        outDir.listFiles()?.forEach { stale ->
            if (stale.isFile && !stale.delete()) {
                Log.w(TAG, "could not delete stale diagnostics file: $stale")
            }
        }
        return outDir
    }

    /**
     * Builds the zip file (exposed package-private so tests can verify its contents without driving
     * the FileProvider). Returns `null` on IO failure.
     */
    @JvmStatic
    fun buildZip(ctx: Context): File? {
        val outDir = prepareOutDir(ctx) ?: return null
        val zipFile = File(outDir, "diagnostics-${System.currentTimeMillis()}.zip")
        val filesDir = ctx.filesDir
        try {
            ZipOutputStream(FileOutputStream(zipFile)).use { zipStream ->
                addBundleDigest(zipStream, filesDir)
                addRecentSessionLogs(zipStream, File(filesDir, SESSION_LOG_DIR))
                addAppLogs(zipStream, File(filesDir, APP_LOG_DIR))
                addSummary(zipStream, File(File(filesDir, SESSION_LOG_DIR), SUMMARY_NAME))
            }
        } catch (ex: IOException) {
            Log.e(TAG, "diagnostics zip failed", ex)
            // Best-effort cleanup so callers don't pick up a half-written zip.
            if (zipFile.exists() && !zipFile.delete()) {
                Log.w(TAG, "could not delete partial zip: $zipFile")
            }
            return null
        }
        return zipFile
    }

    @Throws(IOException::class)
    private fun addBundleDigest(
        zipStream: ZipOutputStream,
        filesDir: File,
    ) {
        // The digest is the AI-facing artifact: already redacted and size-bounded by DiagnosticsBundle,
        // so it ships as the first entry and is written verbatim (no second redaction pass needed).
        val digest = DiagnosticsBundle.build(filesDir)
        val entry =
            ZipEntry(DiagnosticsBundle.BUNDLE_ENTRY_NAME).apply {
                time = System.currentTimeMillis()
            }
        zipStream.putNextEntry(entry)
        zipStream.write(digest.toByteArray(Charsets.UTF_8))
        zipStream.closeEntry()
    }

    @Throws(IOException::class)
    private fun addRecentSessionLogs(
        zipStream: ZipOutputStream,
        sessionLogDir: File,
    ) {
        if (!sessionLogDir.isDirectory) {
            return
        }
        // Filter to the per-session JSONLs (session-*.jsonl). The sessions-summary file is also
        // .jsonl but is added separately via addSummary; including it here would either crowd the
        // 5-file budget or -- worse -- duplicate the entry name in the zip and blow up putNextEntry.
        val files =
            sessionLogDir.listFiles { _, name ->
                name.startsWith("session-") && name.endsWith(".jsonl")
            }
        if (files.isNullOrEmpty()) {
            return
        }
        // Newest first so the slice gives the freshest N files.
        files
            .sortedByDescending { it.lastModified() }
            .take(MAX_SESSION_LOGS)
            .forEach { addFile(zipStream, it, "$SESSION_LOG_DIR/${it.name}", redactText = true) }
    }

    @Throws(IOException::class)
    private fun addAppLogs(
        zipStream: ZipOutputStream,
        appLogDir: File,
    ) {
        if (!appLogDir.isDirectory) {
            return
        }
        appLogDir.listFiles()?.forEach { file ->
            if (file.isFile) {
                addFile(zipStream, file, "$APP_LOG_DIR/${file.name}", redactText = true)
            }
        }
    }

    @Throws(IOException::class)
    private fun addSummary(
        zipStream: ZipOutputStream,
        summary: File,
    ) {
        if (summary.isFile) {
            addFile(zipStream, summary, "$SESSION_LOG_DIR/${summary.name}", redactText = true)
        }
    }

    @Throws(IOException::class)
    private fun addFile(
        zipStream: ZipOutputStream,
        source: File,
        entryName: String,
        redactText: Boolean,
    ) {
        val entry =
            ZipEntry(entryName).apply {
                time = source.lastModified()
            }
        zipStream.putNextEntry(entry)
        if (redactText) {
            val redacted = DataBackup.redactDebugLogText(source.readText(Charsets.UTF_8))
            zipStream.write(redacted.toByteArray(Charsets.UTF_8))
        } else {
            BufferedInputStream(FileInputStream(source)).use { input ->
                val buffer = ByteArray(8192)
                var bytesRead = input.read(buffer)
                while (bytesRead > 0) {
                    zipStream.write(buffer, 0, bytesRead)
                    bytesRead = input.read(buffer)
                }
            }
        }
        zipStream.closeEntry()
    }
}
